### DISCUSSION TITLE
undefined is not iterable (cannot read property Symbol(Symbol.iterator))

### DISCUSSION THREAD

**Discussion by deadcoder0904** [storybookjs/storybook#11756] [Help] [answered]
I have made a React library using https://tsdx.io & chose the React + TypeScript + Storybook template.

The entire code is here → https://github.com/deadcoder0904/react-typical

I get this error:

> undefined is not iterable (cannot read property Symbol(Symbol.iterator))
TypeError: undefined is not iterable (cannot read property Symbol(Symbol.iterator))
    at __read (http://localhost:6006/vendors~main.c9781e3a7458a3b52f4d.bundle.js:177373:46)
    at __spread (http://localhost:6006/vendors~main.c9781e3a7458a3b52f4d.bundle.js:177391:24)
    at http://localhost:6006/main.c9781e3a7458a3b52f4d.bundle.js:77:127
    at commitHookEffectListMount (http://localhost:6006/vendors~main.c9781e3a7458a3b52f4d.bundle.js:140995:26)
    at commitPassiveHookEffects (http://localhost:6006/vendors~main.c9781e3a7458a3b52f4d.bundle.js:141033:11)
    at HTMLUnknownElement.callCallback (http://localhost:6006/vendors~main.c9781e3a7458a3b52f4d.bundle.js:121452:14)
    at Object.invokeGuardedCallbackDev (http://localhost:6006/vendors~main.c9781e3a7458a3b52f4d.bundle.js:121501:16)
    at invokeGuardedCallback (http://localhost:6006/vendors~main.c9781e3a7458a3b52f4d.bundle.js:121556:31)
    at flushPassiveEffectsImpl (http://localhost:6006/vendors~main.c9781e3a7458a3b52f4d.bundle.js:144117:9)
    at unstable_runWithPriority (http://localhost:6006/vendors~main.c9781e3a7458a3b52f4d.bundle.js:170649:12)

Not sure how I can get rid of it?

My storybook file `ReactTypical.stories.tsx` is so simple:

```tsx
import React from 'react';
import { ReactTypical, Props } from '../src';

export default {
  title: 'Basic',
  steps: [
    'Hey',
    5000,
    'you',
    5000,
    'have',
    5000,
    'a',
    5000,
    'blessed',
    5000,
    'day',
  ],
  loop: Infinity,
};

// By passing optional props to this story, you can control the props of the component when
// you consume the story in a test.
export const Default = (props: Props) => <ReactTypical {...props} />;
```

Anyone got any ideas?

**Comment by shilman** [comment]
Fixed your stories:

```js
const props = {
  steps: [
    'Hey',
    5000,
    'you',
    5000,
    'have',
    5000,
    'a',
    5000,
    'blessed',
    5000,
    'day',
  ],
  loop: Infinity,
};
export const Default = () => <ReactTypical {...props} />;
```

**Reply by deadcoder0904** [reply]
Weird, https://tsdx.io has a Storybook template just like the code I wrote in the original comment.

Also, this didn't work. I get the following screen:

![storybook-error](https://user-images.githubusercontent.com/16436270/89120949-2eb72380-d4d8-11ea-9fec-da1c33491f0f.png)

**Reply by shilman** [reply]
~~That template looks wrong to me. Can you file an issue on TSDX?~~

Also, can you see if there's any error on the console log? That updated example worked on my machine.

**Reply by deadcoder0904** [reply]
Nope, no error anywhere even in the console log.

Here's the original storybook template → https://github.com/formium/tsdx/blob/5e5c3f804e68609083be978a080ab905dc7ec19d/templates/react-with-storybook/stories/Thing.stories.tsx which I think would be working

I opened up one yesterday but the maintainer starts blaming lol & locked the conversation https://github.com/formium/tsdx/issues/785

I can open another one but what should I say? I fear the maintainer will again jump up, say the same thing & lock again 🤣

**Reply by deadcoder0904** [reply]
I just used Github's search to find projects using `tsdx` & came across https://github.com/Daarien/react-tsdx-lib/blob/master/stories/Buttons.stories.tsx which uses `tsdx` the same way as shown in `tsdx` + `react` template :)

**Reply by deadcoder0904** [reply]
And there is a console error. Here's the error:

> tslib.es6.js:131 Uncaught TypeError: undefined is not iterable (cannot read property Symbol(Symbol.iterator))
    at __read (tslib.es6.js:131)
    at __spread (tslib.es6.js:149)
    at index.tsx:41
    at commitHookEffectListMount (react-dom.development.js:19731)
    at commitPassiveHookEffects (react-dom.development.js:19769)
    at HTMLUnknownElement.callCallback (react-dom.development.js:188)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:237)
    at invokeGuardedCallback (react-dom.development.js:292)
    at flushPassiveEffectsImpl (react-dom.development.js:22853)
    at unstable_runWithPriority (scheduler.development.js:653)
__read @ tslib.es6.js:131
__spread @ tslib.es6.js:149
(anonymous) @ index.tsx:41
commitHookEffectListMount @ react-dom.development.js:19731
commitPassiveHookEffects @ react-dom.development.js:19769
callCallback @ react-dom.development.js:188
invokeGuardedCallbackDev @ react-dom.development.js:237
invokeGuardedCallback @ react-dom.development.js:292
flushPassiveEffectsImpl @ react-dom.development.js:22853
unstable_runWithPriority @ scheduler.development.js:653
runWithPriority$1 @ react-dom.development.js:11039
flushPassiveEffects @ react-dom.development.js:22820
(anonymous) @ react-dom.development.js:22699
workLoop @ scheduler.development.js:597
flushWork @ scheduler.development.js:552
performWorkUntilDeadline @ scheduler.development.js:164

> The above error occurred in the <Typical> component:
    in Typical
    in Unknown (created by Context.Consumer)
    in InlineZoomWrapper (created by Context.Consumer)
    in div (created by InlineStory)
    in InlineStory (created by Story)
    in Story (created by Context.Consumer)
    in MDXProvider (created by Context.Consumer)
    in div (created by Context.Consumer)
    in StoryContainer (created by DocsStory)
    in div (created by Context.Consumer)
    in Styled(div) (created by Preview)
    in div (created by Context.Consumer)
    in Styled(div) (created by Preview)
    in div (created by Context.Consumer)
    in Styled(div) (created by Preview)
    in div (created by Context.Consumer)
    in Styled(div) (created by Preview)
    in Preview (created by Context.Consumer)
    in Preview (created by DocsStory)
    in div (created by Anchor)
    in Anchor (created by DocsStory)
    in DocsStory (created by Primary)
    in Primary (created by DocsPage)
    in DocsPage
    in div (created by Context.Consumer)
    in Styled(div) (created by DocsContainer)
    in div (created by Context.Consumer)
    in Styled(div) (created by DocsContainer)
    in MDXProvider (created by DocsContainer)
    in ThemeProvider (created by DocsContainer)
    in DocsContainer

> Consider adding an error boundary to your tree to customize error handling behavior.
Visit https://fb.me/react-error-boundaries to learn more about error boundaries.
logCapturedError @ react-dom.development.js:19527
logError @ react-dom.development.js:19564
update.callback @ react-dom.development.js:20708
callCallback @ react-dom.development.js:12490
commitUpdateQueue @ react-dom.development.js:12511
commitLifeCycles @ react-dom.development.js:19883
commitLayoutEffects @ react-dom.development.js:22803
callCallback @ react-dom.development.js:188
invokeGuardedCallbackDev @ react-dom.development.js:237
invokeGuardedCallback @ react-dom.development.js:292
commitRootImpl @ react-dom.development.js:22541
unstable_runWithPriority @ scheduler.development.js:653
runWithPriority$1 @ react-dom.development.js:11039
commitRoot @ react-dom.development.js:22381
finishSyncRender @ react-dom.development.js:21807
performSyncWorkOnRoot @ react-dom.development.js:21793
(anonymous) @ react-dom.development.js:11089
unstable_runWithPriority @ scheduler.development.js:653
runWithPriority$1 @ react-dom.development.js:11039
flushSyncCallbackQueueImpl @ react-dom.development.js:11084
flushSyncCallbackQueue @ react-dom.development.js:11072
flushPassiveEffectsImpl @ react-dom.development.js:22883
unstable_runWithPriority @ scheduler.development.js:653
runWithPriority$1 @ react-dom.development.js:11039
flushPassiveEffects @ react-dom.development.js:22820
(anonymous) @ react-dom.development.js:22699
workLoop @ scheduler.development.js:597
flushWork @ scheduler.development.js:552
performWorkUntilDeadline @ scheduler.development.js:164

> tslib.es6.js:131 Uncaught TypeError: undefined is not iterable (cannot read property Symbol(Symbol.iterator))
    at __read (tslib.es6.js:131)
    at __spread (tslib.es6.js:149)
    at index.tsx:41
    at commitHookEffectListMount (react-dom.development.js:19731)
    at commitPassiveHookEffects (react-dom.development.js:19769)
    at HTMLUnknownElement.callCallback (react-dom.development.js:188)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:237)
    at invokeGuardedCallback (react-dom.development.js:292)
    at flushPassiveEffectsImpl (react-dom.development.js:22853)
    at unstable_runWithPriority (scheduler.development.js:653)
__read @ tslib.es6.js:131
__spread @ tslib.es6.js:149
(anonymous) @ index.tsx:41
commitHookEffectListMount @ react-dom.development.js:19731
commitPassiveHookEffects @ react-dom.development.js:19769
callCallback @ react-dom.development.js:188
invokeGuardedCallbackDev @ react-dom.development.js:237
invokeGuardedCallback @ react-dom.development.js:292
flushPassiveEffectsImpl @ react-dom.development.js:22853
unstable_runWithPriority @ scheduler.development.js:653
runWithPriority$1 @ react-dom.development.js:11039
flushPassiveEffects @ react-dom.development.js:22820
(anonymous) @ react-dom.development.js:22699
workLoop @ scheduler.development.js:597
flushWork @ scheduler.development.js:552
performWorkUntilDeadline @ scheduler.development.js:164

It looks like it's around my code at `index.tsx:41` which is around `...steps`:

```tsx
type(typicalRef.current as HTMLElement, ...steps);
```

But the code works & the app is also working, only storybook doesn't work.

Not sure how to solve this?

**Reply by shilman** [reply]
The tsdx template looks fine. The part that's buggy is the part you added.

**Reply by deadcoder0904** [reply]
Yeah, any ideas how to solve it? I can't seem to find an answer anywhere sadly :(

**Reply by kylemh** [reply]
You may need to check that the ref exists in your hook. I’d also recommend accepting a ref as a prop instead as part of your component.

**Reply by deadcoder0904** [reply]
Made some progress.

I tried putting `console.log` in the code:

```tsx
React.useEffect(() => {
    console.log(`React.useEffect`, steps);
    if (loop === Infinity) {
      console.log(`if`);
      type(typicalRef.current as HTMLElement, ...steps, loopedType);
    } else if (typeof loop === 'number') {
      console.log(`elseif`);
      type(
        typicalRef.current as HTMLElement,
        ...Array(loop)
          .fill(steps)
          .flat()
      );
    } else {
      console.log(`else`);
      type(typicalRef.current as HTMLElement, ...steps);
    }
  }, [typicalRef, loop, steps]);
```

And it logged when I run the `examples/` folder:

```bash
React.useEffect ["Creator", 5000, "Entrepreneur", 5000, "Doer", 5000, "Lone Wolf", 5000]
if
```

But if I do run it from the storybook, it logs the following:

```bash
React.useEffect undefined
else
```

Now I'm pretty sure the `steps` is undefined in Storybook, it's definitely not a ref issue. Idk how `steps` is not accessible in Storybook even though I pass it 🤔

Even another prop `loop` isn't getting passed.

Here's my storybook code:

```tsx
import React from 'react';
import { ReactTypical, Props } from '../src';

export default {
  title: 'Basic',
  steps: [
    'Hey',
    5000,
    'you',
    5000,
    'have',
    5000,
    'a',
    5000,
    'blessed',
    5000,
    'day',
  ],
  loop: Infinity,
};

// By passing optional props to this story, you can control the props of the component when
// you consume the story in a test.
export const Default = (props: Props) => {
  console.log(props);
  return <ReactTypical {...props} />;
};
```

**Comment by deadcoder0904** [comment, marked as answer]
Alright I solved it like this:

```tsx
import React from 'react';
import { ReactTypical, Props } from '../src';

export default {
  title: 'Basic',
};

export const Default = (props: Props) => {
  return (
    <ReactTypical
      {...props}
      steps={[
        'Hey',
        500,
        'you',
        500,
        'have',
        500,
        'a',
        500,
        'blessed',
        500,
        'day',
      ]}
    />
  );
};
```

**Reply by kylemh** [reply]
@shilman are people supposed to be able to pass any/all arguments in the default export? My understanding was that it's just reserved for Storybook metadata.

**Reply by shilman** [reply]
@kylemh no, he made that up

**Reply by kylemh** [reply]
If Storybook doesn't have a lint plugin, this could be a good reason for one. I've seen other people put whatever they want in the default export.

**Reply by shilman** [reply]
@kylemh related #11032

**Reply by deadcoder0904** [reply]
> @kylemh no, he made that up

@kylemh @shilman can't blame me, the `tsdx` had an example just like that so I just tweaked it to my needs :)

**Reply by kylemh** [reply]
@deadcoder0904 no it didn't. [I made the example/template](https://github.com/formium/tsdx/blob/master/templates/react-with-storybook/stories/Thing.stories.tsx).

To be clear, I don't doubt you found it somewhere! I want to find it so I can delete it - nobody should have to deal with what you dealt with.

**Reply by deadcoder0904** [reply]
@kylemh so yeah I tweaked that thing only. But I think the example could be more advanced. I thought the `title` is a prop as I have never used Storybook before.

Just making the example with 1 more prop would be more clear than that template.

As I can see `Thing` doesn't accept any props so maybe make a little simpler component that accepts a prop like `name`:

```tsx
const Greeting = ({ name }) => return `Hello ${name}`
```

So the Storybook example can become the one with props :)

**Reply by kylemh** [reply]
Right, that's what we're trying to explain here. `title` wasn't a prop for `Thing` in that example.

The default export for a [Component Story Format (CSF)](https://storybook.js.org/docs/formats/component-story-format/) file does not have keys which represent props - they're strictly reserved for specific keywords.

Namely...
- title
- component
- decorators
- parameters
- includeStories
- excludeStories

**Reply by deadcoder0904** [reply]
Yeah, that I understand now, but it just wasn't clear to a beginner me yesterday so I'm suggesting an example with 1 prop for other beginners :)
